frappe/public/js/frappe/form/form.js

```diff
@@ -421,6 +422,17 @@ frappe.ui.form.Form = class FrappeForm {
 		}
 	}
 
+
+	// Adds un beforeunload Listener to trigger a "Leave Site" Dialog, when a dirty (unsaved) form is closed
+	set_dirty_listener() {
+		addEventListener('beforeunload', (event) => {
+			if (cur_frm.is_dirty()) {
+			event.preventDefault();
+			return event.returnValue = 'There are unsaved changes, are you sure you want to exit?';
+			}
```

Comment:
Few concerns:

1. This should probably be disabled in developer mode (annoying with JS build reloader) `frappe.boot.developer_mode` flag can be used. 
2. The return value doesn't appear in chrome at least, I got default warning `Changes that you made may not be saved.`
3. Check if `cur_frm` is truthy before checking if it's dirty, it might be unset if user moved to another page. 

<img width="1072" alt="Screenshot 2022-06-02 at 3 15 26 PM" src="https://user-images.githubusercontent.com/9079960/171604014-99e1eb69-86bd-4322-a8c0-f3821f2dbd61.png">